tardis/model/base.py

```diff
@@ -0,0 +1,186 @@
+            self._velocity[-1] = self.v_boundary_outer
+        return self._velocity
+
+    @property
+    def v_inner(self):
+        return self.velocity[:-1]
+
+    @property
+    def v_outer(self):
+        return self.velocity[1:]
+
+    @property
+    def v_middle(self):
+        return 0.5 * self.v_inner + 0.5 * self.v_outer
+
+    @property
+    def density(self):
+        return self.homologous_density
+
+    @property
+    def volume(self):
+        return ((4. / 3) * np.pi * (self.r_outer ** 3 - self.r_inner ** 3)).cgs
+
+    @property
+    def no_of_shells(self):
+        return len(self.velocity) - 1
+
+    @property
+    def v_boundary_inner(self):
+        if self._v_boundary_inner is None:
+            return self.raw_velocity[0]
+        return self._v_boundary_inner
+
+    @v_boundary_inner.setter
+    def v_boundary_inner(self, value):
+        if value is not None:
+            if value > self.v_boundary_outer:
+                raise ValueError('v_boundary_inner must not be higher than '
+                                 'v_boundary_outer.')
+            if value > self.raw_velocity[-1]:
+                raise ValueError('v_boundary_inner is outside of '
+                                 'the model range.')
+            value = max(value, self.raw_velocity[0])
+        self._v_boundary_inner = value
+        # Invalidate the cached cut-down velocity array
+        self._velocity = None
+
+    @property
+    def v_boundary_outer(self):
+        if self._v_boundary_outer is None:
+            return self.raw_velocity[-1]
+        return self._v_boundary_outer
+
+    @v_boundary_outer.setter
+    def v_boundary_outer(self, value):
+        if value is not None:
+            if value < self.v_boundary_inner:
+                raise ValueError('v_boundary_outer must not be smaller than '
+                                 'v_boundary_inner.')
+            if value < self.raw_velocity[0]:
+                raise ValueError('v_boundary_outer is outside of '
+                                 'the model range.')
+            value = min(value, self.raw_velocity[-1])
+        self._v_boundary_outer = value
+        # Invalidate the cached cut-down velocity array
+        self._velocity = None
+
+    @property
+    def v_boundary_inner_index(self):
+        if self.v_boundary_inner <= self.raw_velocity[0]:
+            return None
+        else:
+            idx = max(0,
+                      self.raw_velocity.searchsorted(self.v_boundary_inner) - 1)
+            # check for zero volume of designated first cell
+            if np.isclose(self.v_boundary_inner, self.raw_velocity[idx + 1],
+                          atol=1e-8 * u.km / u.s) and (self.v_boundary_inner <=
+                                                           self.raw_velocity[idx + 1]):
+                idx += 1
+            return idx
+
+    @property
+    def v_boundary_outer_index(self):
+        if self.v_boundary_outer >= self.raw_velocity[-1]:
```

Comment:
you are extrapolating here. both boundaries should be within the model and inner < outer. I think we should have a function where we have all these rules in place.